liquidhaskell-boot/src/Language/Haskell/Liquid/Bare/Axiom.hs

```diff
@@ -42,6 +62,88 @@ makeHaskellAxioms cfg src env tycEnv name lmap spSig spec = do
+                       -> (LocSymbol, LocSymbol) -- actual function and pretended function
+                       -> (Ghc.Var, LocSpecType, F.Equation)
+-----------------------------------------------------------------------------------------------
+makeAssumeReflectAxiom sig env tce name (actual, pretended) =
+  -- The actual and pretended function must have the same type
+  if pretendedTy == actualTy then
+    (actualV, actual {val = aty_at `strengthenRes` ref} , actualEq)
+  else
+    Ex.throw $ mkError actual $
+      show qPretended ++ " and " ++ show qActual ++ " should have the same type. But " ++
+      "types " ++ F.showpp pretendedTy ++ " and " ++ F.showpp actualTy  ++ " do not match."
+  where
+    -- Get the Ghc.Var's of the actual and pretended function names
+    actualV = case Bare.lookupGhcVar env name "wiredAxioms" actual of
+      Right x -> x
+      Left _ -> Ex.throw $ mkError actual $ "Not in scope: " ++ show (val actual)
+    pretendedV = case Bare.lookupGhcVar env name "wiredAxioms" pretended of
+      Right x -> x
+      Left _ -> Ex.throw $ mkError pretended $ "Not in scope: " ++ show (val pretended)
+    -- Get the qualified name symbols for the actual and pretended functions
+    qActual = Bare.qualifyTop env name (F.loc actual) (val actual)
+    qPretended = Bare.qualifyTop env name (F.loc pretended) (val pretended)
+    -- Get the GHC type of the actual and pretended functions
+    actualTy = Ghc.varType actualV
+    pretendedTy = Ghc.varType pretendedV
+    -- Compute argument names for the actual/pretended functions
+    -- The argument names are lq1, lq2, etc.
+    -- These argument names will be used in the equation
+    args = getArgs 1 actualTy
+    -- Function types can be of multiple sorts. We are only interested in the Type -> Type (i.e., ->) ones,
+    -- which correspond to flag Ghc.FTF_T_T. For (=>) Constraint -> Type (flag FTF_C_T), we just ignore the
+    -- (constraint) argument. For the others kinds, we throw are error, as they are unsupported.
+    getArgs n Ghc.FunTy{ft_arg=ty0, ft_res=ty1, ft_af=Ghc.FTF_T_T} = (F.symbol . ("lq" ++) . show $ n, typeSort tce ty0) : getArgs (n+1) ty1
+    getArgs n Ghc.FunTy{ft_res=ty1, ft_af=Ghc.FTF_C_T} = getArgs n ty1
+    getArgs _ Ghc.FunTy{} = Ex.throw $ mkError actual "This function cannot be `assume reflect`'ed. Its signature is too complex for me."
+    getArgs n (Ghc.ForAllTy _ ty) = getArgs n ty
+    getArgs _ _ = []
+
+    -- Compute the refined type of the actual function. See `makeAssumeType` for details
+    at    = axiomType allowTC actual rt
+    aty_at = aty at
+    out   = rTypeSort tce $ ares at
+    mbT   = val <$> lookup actualV sigs
+    allowTC = typeclass (getConfig env)
+    sigs                    = gsTySigs sig
+    rt    = fromRTypeRep .
+            (\trep@RTypeRep{..} ->
+                trep{ty_info = fmap (\i -> i{permitTC = Just allowTC}) ty_info}) .
+            toRTypeRep $ Mb.fromMaybe (ofType actualTy) mbT
```

Comment:
This part is a bit obscure.

Reply:
I can only agree, but actually that's just a reuse of the code of `makeAssumeType`, which I also wished was commented more. Hard to comment it in detail w/o risking of commenting it incorrectly